From: Shiz <hi@shiz.me>
Updated by Rasmus Thomsen <oss@cogitri.dev> on 28th of July 2019
Updated again by Ariadne Conill <ariadne@dereferenced.org> on 31 October 2020
Date: Thu, 20 Aug 2017 01:52:36 +0200
Last-Updated: Sat, 28 Oct 2017 20:23:00 +0200
Subject: [PATCH] Add openwrt targets

This adds `$arch-open-linux-musl` targets to Rust to encode our toolchain
and distribution-specific quirks instead of polluting the main musl target of
`$arch-unknown-linux-musl`.

--- /dev/null
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/aarch64_openwrt_linux_musl.rs
@@ -0,0 +1,13 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::aarch64_unknown_linux_musl::target()?;
+
+    base.llvm_target = "aarch64-openwrt-linux-musl".to_string();
+    base.target_vendor = "openwrt".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+
+
+    Ok(base)
+}
--- /dev/null
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/armv6_openwrt_linux_musleabihf.rs
@@ -0,0 +1,13 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::arm_unknown_linux_musleabihf::target()?;
+
+    base.llvm_target = "armv6-openwrt-linux-musleabihf".to_string();
+    base.target_vendor = "openwrt".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+
+
+    Ok(base)
+}
--- /dev/null
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/armv7_openwrt_linux_musleabihf.rs
@@ -0,0 +1,13 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::armv7_unknown_linux_musleabihf::target()?;
+
+    base.llvm_target = "armv7-openwrt-linux-musleabihf".to_string();
+    base.target_vendor = "openwrt".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+
+
+    Ok(base)
+}
--- /dev/null
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/i586_openwrt_linux_musl.rs
@@ -0,0 +1,14 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::i686_unknown_linux_musl::target()?;
+
+    base.options.cpu = "pentium4".to_string();
+    base.llvm_target = "i586-openwrt-linux-musl".to_string();
+    base.target_vendor = "openwrt".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+
+
+    Ok(base)
+}
--- /dev/null
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/mips64_openwrt_linux_musl.rs
@@ -0,0 +1,13 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::mips64_unknown_linux_muslabi64::target()?;
+
+    base.llvm_target = "mips64-openwrt-linux-musl".to_string();
+    base.target_vendor = "openwrt".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+
+
+    Ok(base)
+}
--- /dev/null
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/mips64el_openwrt_linux_musl.rs
@@ -0,0 +1,13 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::mips64el_unknown_linux_muslabi64::target()?;
+
+    base.llvm_target = "mips64el-openwrt-linux-musl".to_string();
+    base.target_vendor = "openwrt".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+
+
+    Ok(base)
+}
--- /dev/null
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/mips_openwrt_linux_musl.rs
@@ -0,0 +1,13 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::mips_unknown_linux_musl::target()?;
+
+    base.llvm_target = "mips-openwrt-linux-musl".to_string();
+    base.target_vendor = "openwrt".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+
+
+    Ok(base)
+}
--- /dev/null
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/mipsel_openwrt_linux_musl.rs
@@ -0,0 +1,13 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::mipsel_unknown_linux_musl::target()?;
+
+    base.llvm_target = "mipsel-openwrt-linux-musl".to_string();
+    base.target_vendor = "openwrt".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+
+
+    Ok(base)
+}
--- rustc-1.47.0-src.orig/compiler/rustc_target/src/spec/mod.rs
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/mod.rs
@@ -490,6 +490,19 @@
 }
 
 supported_targets! {
+    ("i586-openwrt-linux-musl", i586_openwrt_linux_musl),
+    ("x86_64-openwrt-linux-musl", x86_64_openwrt_linux_musl),
+    ("aarch64-openwrt-linux-musl", aarch64_openwrt_linux_musl),
+    ("armv6-openwrt-linux-musleabihf", armv6_openwrt_linux_musleabihf),
+    ("armv7-openwrt-linux-musleabihf", armv7_openwrt_linux_musleabihf),
+    ("powerpc-openwrt-linux-musl", powerpc_openwrt_linux_musl),
+    ("powerpc64-openwrt-linux-musl", powerpc64_openwrt_linux_musl),
+    ("powerpc64le-openwrt-linux-musl", powerpc64le_openwrt_linux_musl),
+    ("s390x-openwrt-linux-musl", s390x_openwrt_linux_musl),
+    ("mips-openwrt-linux-musl", mips_openwrt_linux_musl),
+    ("mipsel-openwrt-linux-musl", mipsel_openwrt_linux_musl),
+    ("mips64-openwrt-linux-musl", mips64_openwrt_linux_musl),
+    ("mips64el-openwrt-linux-musl", mips64el_openwrt_linux_musl),
     ("x86_64-unknown-linux-gnu", x86_64_unknown_linux_gnu),
     ("x86_64-unknown-linux-gnux32", x86_64_unknown_linux_gnux32),
     ("i686-unknown-linux-gnu", i686_unknown_linux_gnu),
--- /dev/null
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/powerpc64_openwrt_linux_musl.rs
@@ -0,0 +1,13 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::powerpc64_unknown_linux_musl::target()?;
+
+    base.llvm_target = "powerpc64-openwrt-linux-musl".to_string();
+    base.target_vendor = "openwrt".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+
+
+    Ok(base)
+}
--- /dev/null
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/powerpc64le_openwrt_linux_musl.rs
@@ -0,0 +1,13 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::powerpc64le_unknown_linux_musl::target()?;
+
+    base.llvm_target = "powerpc64le-openwrt-linux-musl".to_string();
+    base.target_vendor = "openwrt".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+
+
+    Ok(base)
+}
--- /dev/null
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/powerpc_openwrt_linux_musl.rs
@@ -0,0 +1,13 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::powerpc_unknown_linux_musl::target()?;
+
+    base.llvm_target = "powerpc-openwrt-linux-musl".to_string();
+    base.target_vendor = "openwrt".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+
+
+    Ok(base)
+}
--- /dev/null
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/s390x_openwrt_linux_musl.rs
@@ -0,0 +1,13 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::s390x_unknown_linux_musl::target()?;
+
+    base.llvm_target = "s390x-openwrt-linux-musl".to_string();
+    base.target_vendor = "openwrt".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+
+
+    Ok(base)
+}
--- /dev/null
+++ rustc-1.47.0-src/compiler/rustc_target/src/spec/x86_64_openwrt_linux_musl.rs
@@ -0,0 +1,13 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::x86_64_unknown_linux_musl::target()?;
+
+    base.llvm_target = "x86_64-openwrt-linux-musl".to_string();
+    base.target_vendor = "openwrt".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+
+
+    Ok(base)
+}
